xla/service/gpu/double_buffer_loop_unrolling.cc

```diff
@@ -83,6 +83,169 @@ void SetChannelIdForNewCollective(HloInstruction* new_instr,
   }
 }
 
+absl::StatusOr<std::vector<std::pair<int64_t, int64_t>>> ParseVectorOfPairs(
+    std::basic_string<char>& str) {
+  TF_ASSIGN_OR_RETURN(std::vector<ReplicaGroup> replica_groups,
+                      ParseReplicaGroupsOnly(str));
+  std::vector<std::pair<int64_t, int64_t>> res;
+  res.reserve(replica_groups.size());
+  for (const ReplicaGroup& replica_group : replica_groups) {
+    TF_RET_CHECK(replica_group.replica_ids_size() == 2);
+    int64_t a = replica_group.replica_ids(0);
+    int64_t b = replica_group.replica_ids(1);
+    res.emplace_back(a, b);
+  }
+  return res;
+}
+
+absl::Status SetSendRecvValidationForPeeledInstr(HloInstruction* new_instr,
+                                                 HloInstruction* old_instr) {
+  assert(new_instr->opcode() == old_instr->opcode() &&
```

Comment:
We do not use `assert` inside, either CHECK if you want to crash the whole process (very rare) or TF_RET_CHECK to return Status.